tests/functional/test_application_restricted.py

```diff
@@ -180,94 +179,100 @@ def auth():
 
 @given("I create a new app")
 def create_test_app(setup_session, context):
-    product, app, token = setup_session
+    product, app, token_response, developer_apps, api_products = setup_session
     context["product"] = product
     context["app"] = app
-    context["token"] = token
+    context["token_response"] = token_response
+    context["developer_apps"] = developer_apps
+    context["api_products"] = api_products
 
 
 @given(parsers.parse("I add the attribute with key of {key} and a value of {value}"))
 def add_custom_attribute_to_app(key: str, value: str, context: dict):
 
     app = context["app"]
-
-    asyncio.run(
-        app.set_custom_attributes(
-            {"jwks-resource-url": config.JWKS_RESOURCE_URL, key: value}
-        )
+    developer_apps = context["developer_apps"]
+    app.set_custom_attributes(
+        {"jwks-resource-url": config.JWKS_RESOURCE_URL, key: value},
+        developer_apps
     )
 
 
 @given("I add an asid attribute")
 def add_asid_attribute_to_app(context: dict):
 
     app = context["app"]
-
-    asyncio.run(
-        app.set_custom_attributes(
-            {
-                "jwks-resource-url": config.JWKS_RESOURCE_URL,
-                "asid": config.ENV["internal_dev_asid"],
-            }
-        )
+    developer_apps = context["developer_apps"]
+    app.set_custom_attributes(
+        {
+            "jwks-resource-url": config.JWKS_RESOURCE_URL,
+            "asid": config.ENV["internal_dev_asid"],
+        },
+        developer_apps
     )
 
 
 @given(parsers.parse("I add the scope {scope}"))
 def add_scope_to_product(scope, context):
     product = context["product"]
-    asyncio.run(product.update_scopes([scope]))
+    api_products = context["api_products"]
+    product.update_scopes([scope], api_products)
 
 
 @given("I have a valid access token")
 def set_valid_access_token(auth, context):
 
-    claims = {
-        "sub": config.APPLICATION_RESTRICTED_API_KEY,
-        "iss": config.APPLICATION_RESTRICTED_API_KEY,
-        "jti": str(uuid.uuid4()),
-        "aud": f"{config.BASE_URL}/oauth2/token",
-        "exp": int(time.time()) + 300,
-    }
+    token_response = context.get("token_response")
 
-    # If one exists, use the client_id of the test app instead
-    if "app" in context:
-        claims.update(
-            {
-                "sub": context["app"].get_client_id(),
-                "iss": context["app"].get_client_id(),
-            }
-        )
+    # If a token is not available in the context, get a new one
+    if not token_response:
+        claims = {
+            "sub": config.APPLICATION_RESTRICTED_API_KEY,
+            "iss": config.APPLICATION_RESTRICTED_API_KEY,
+            "jti": str(uuid.uuid4()),
+            "aud": f"{config.BASE_URL}/{config.OAUTH_PROXY}/token",
+            "exp": int(time.time()) + 300,
+        }
```

Comment:
This feels like we are holding on to old patterns while trying to implement new ones, as well. Should we call for a new token through the pytest-nhsd-apim fixtures, similar to the one that would/should be in the context? Or is this another, add it to the refactor ticket?